From 79b99f7d65cce0730998a84d2fb615ae5ad0cff5 Mon Sep 17 00:00:00 2001
From: Andy Green <andy@openmoko.com>
Date: Wed, 2 Jul 2008 22:39:01 +0100
Subject: [PATCH] debug-codec-register-setting-sysfs.patch
 Allow direct setting of codec registers for super mega meddling power
 Signed-off-by: Andy Green <andy@openmoko.com>

---
 sound/soc/soc-core.c |   35 +++++++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index 25f7818..9e20333 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -928,6 +928,38 @@ static ssize_t codec_reg_show(struct device *dev,
 }
 static DEVICE_ATTR(codec_reg, 0444, codec_reg_show, NULL);
 
+
+static ssize_t codec_reg_write(struct device *dev,
+			       struct device_attribute *attr,
+			       const char *buf, size_t count)
+{
+	u32 address;
+	u32 data;
+	char * end;
+	size_t left = count;
+	struct snd_soc_device *devdata = dev_get_drvdata(dev);
+	struct snd_soc_codec *codec = devdata->codec;
+
+	address = simple_strtoul(buf, &end, 16);
+	left -= (int)(end - buf);
+	while ((*end == ' ') && (left)) {
+		end++;
+		left--;
+	}
+	if (!left)
+		return count;
+	data = simple_strtoul(end, &end, 16);
+
+	printk(KERN_INFO"user writes Codec reg 0x%02X with Data 0x%04X\n",
+								address, data);
+
+	codec->write(codec, address, data);
+
+	return count;
+}
+
+static DEVICE_ATTR(codec_reg_write, 0644, NULL, codec_reg_write);
+
 /**
  * snd_soc_new_ac97_codec - initailise AC97 device
  * @codec: audio codec
@@ -1139,6 +1171,9 @@ int snd_soc_register_card(struct snd_soc_device *socdev)
 	err = device_create_file(socdev->dev, &dev_attr_codec_reg);
 	if (err < 0)
 		printk(KERN_WARNING "asoc: failed to add codec sysfs entries\n");
+	err = device_create_file(socdev->dev, &dev_attr_codec_reg_write);
+	if (err < 0)
+		printk(KERN_WARNING "asoc: failed to add codec sysfs entries\n");
 out:
 	mutex_unlock(&codec->mutex);
 	return ret;
-- 
1.5.6.5

